x/stake/cross_stake/cross_stake.go

```diff
@@ -0,0 +1,364 @@
+		ackBytes, err := rlp.EncodeToBytes(ackPack)
+		if err != nil {
+			return sdk.ExecuteResult{}, err
+		}
+		return sdk.ExecuteResult{
+			Err:     sdkErr,
+			Payload: ackBytes,
+		}, nil
+	}
+
+	delAddr := types.GetStakeCAoB(pack.DelAddr[:], types.DelegateCAoBSalt)
+	shares, sdkErr := app.stakeKeeper.ValidateUnbondAmount(ctx, delAddr, pack.ValSrc, pack.Amount.Int64())
+	if sdkErr != nil {
+		ackPack := types.NewCrossStakeRedelegationAckPackage(pack, types.CrossStakeTypeRedelegate, CrossStakeErrBadDelegation)
+		ackBytes, err := rlp.EncodeToBytes(ackPack)
+		if err != nil {
+			return sdk.ExecuteResult{}, err
+		}
+		return sdk.ExecuteResult{
+			Err:     sdkErr,
+			Payload: ackBytes,
+		}, nil
+	}
+
+	_, err := app.stakeKeeper.BeginRedelegation(ctx.WithCrossStake(true), delAddr, pack.ValSrc, pack.ValDst, shares)
+	if err != nil {
+		return sdk.ExecuteResult{}, err
+	}
+
+	// publish redelegate event
+	if app.stakeKeeper.PbsbServer != nil && ctx.IsDeliverTx() {
+		event := types.SideRedelegateEvent{
+			RedelegateEvent: types.RedelegateEvent{
+				StakeEvent: types.StakeEvent{
+					IsFromTx: true,
+				},
+				Delegator:    delAddr,
+				SrcValidator: pack.ValSrc,
+				DstValidator: pack.ValDst,
+				Amount:       pack.Amount.Int64(),
+				Denom:        app.stakeKeeper.BondDenom(ctx),
+				TxHash:       ctx.Value(baseapp.TxHashKey).(string),
+			},
+			SideChainId: sideChainId,
+		}
+		app.stakeKeeper.PbsbServer.Publish(event)
+	}
+
+	resultTags := sdk.NewTags(
+		types.TagCrossStakePackageType, []byte(uint8(types.CrossStakeTypeRedelegate)),
+	)
+	return sdk.ExecuteResult{
+		Tags: resultTags,
+	}, nil
+}
+
+func (app *CrossStakeApp) handleDistributeRewardRefund(ctx sdk.Context, pack *types.CrossStakeRefundPackage) (sdk.ExecuteResult, error) {
+	symbol := app.stakeKeeper.BondDenom(ctx)
+	refundAmount := bsc.ConvertBSCAmountToBCAmount(pack.Amount)
+	coins := sdk.Coins{sdk.NewCoin(symbol, refundAmount)}
+	delAddr := types.GetStakeCAoB(pack.Recipient[:], types.DelegateCAoBSalt)
+	refundAddr := types.GetStakeCAoB(delAddr.Bytes(), types.RewardCAoBSalt)
+	_, err := app.stakeKeeper.BankKeeper.SendCoins(ctx, sdk.PegAccount, refundAddr, coins)
+	if err != nil {
+		return sdk.ExecuteResult{}, err
+	}
+
+	// publish  event
+	if app.stakeKeeper.PbsbServer != nil && ctx.IsDeliverTx() {
+		app.stakeKeeper.AddrPool.AddAddrs([]sdk.AccAddress{sdk.PegAccount, refundAddr})
+		PublishCrossStakeEvent(ctx, app.stakeKeeper, sdk.PegAccount.String(), []types.CrossReceiver{{refundAddr.String(), pack.Amount.Int64()}},
+			app.stakeKeeper.BondDenom(ctx), types.CrossStakeDistributeRewardFailAckRefundType, 0)
+	}
+
+	return sdk.ExecuteResult{
+		Tags: sdk.Tags{sdk.GetPegOutTag(symbol, refundAmount)},
+	}, nil
+}
+
+func (app *CrossStakeApp) handleDistributeUndelegatedRefund(ctx sdk.Context, pack *types.CrossStakeRefundPackage) (sdk.ExecuteResult, error) {
+	symbol := app.stakeKeeper.BondDenom(ctx)
+	refundAmount := bsc.ConvertBSCAmountToBCAmount(pack.Amount)
```

Comment:
Need move outside this function and only needed for `failAck` packages.

Reply:
Done